feat: script for downloading sourcemaps + symbolicating them

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR adds a script that can be used to symbolicate profiles recorded from a production (staging) app.

To symbolicate a profile we need a source map. We need the exact source map that was used when the production app was created. 
Thus we upload the source maps during the deploy process in the GitHub action workflow as artifact.

The now added script in `scripts/symbolicate-profile.ts`:

- takes a `.cpuprofile` file and gets the app version from it (we encode it in the filename)
- downloads the source map from the GitHub artifacts
	1. It gets the artifact from the workflow runs by the file name
	2. It creates a download link for the artifact
	3. It downloads it as zip and unpacks it
	4. It stores the source map in a `./.sourcemap` folder. Next time the command is run it tries to reuse the source map
- symbolicates the profile using react-native-release-profiler CLI

The script can simply be invoked by running:

```bash
npm run symbolicate-profile -- --platform=android --profile='./Profile_trace_for_1.4.86-0.cpuprofile'
```

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
8. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/43649
PROPOSAL: https://github.com/Expensify/App/issues/43649#issuecomment-2172557088


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
9. Verify a modal appears displaying a preview of that image
--->

No QA, as its a dev only change.
For PR review I'd appreciate if someone else could test the script.

You can use this profile for testing:
[Profile_trace_for_1.4.86-0.cpuprofile](https://github.com/user-attachments/files/15911389/Profile_trace_for_1.4.86-0.cpuprofile)


### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

n/a - dev only change 

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.


### Screenshots/Videos

https://github.com/Expensify/App/assets/16821682/73091c1d-4a2a-4101-b7fd-8758ff81f5f7

hannojg committed:
feat: script for getting uploaded sourcemaps

hannojg committed:
finish implementation of download source map step

hannojg committed:
symbolicate

hannojg committed:
fix help command + lint

hannojg committed:
lint

hannojg committed:
fix bugs

Julesssss commented:
Hey @hannojg, can we hold this PR for a moment? There's a chance that [part 1](https://github.com/Expensify/App/pull/43823) of these changes broke the iOS build

hannojg commented:
Sure thing! Can I help with the broken iOS job?

Julesssss commented:
> Sure thing! Can I help with the broken iOS job?

Just saw this, build is okay now after the revert 👍

hannojg commented:
Actually, this PR can already be tested and merged even with the first PR being reverted. The script won't really work with no source maps uploaded, but still (or we could also wait, I don't mind)

hannojg commented:
Oh, sorry, I just found a way to make the querying for the artifact way simpler. Will move this too draft until its changed and ready.

hannojg committed:
simplify fetching artifact by using artifact name

hannojg committed:
run prettier

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+
+const argsMap = parseCommandLineArguments();
+
+// #region Input validation
```
What's with the `#region` / `#endregion` comments? Never seen these before.

imho if you want to make the sections of the script clearer, use comments in this style:

```ts
/* ============== INPUT VALIDATION ============== */
```

or even like this:

```ts
/* ==============================================
   ============== INPUT VALIDATION ============== 
   ============================================== */
```

roryabraham replied:
I guess it's a VSCode thing 🤷🏼

roryabraham commented on src/components/ProfilingToolMenu/BaseProfilingToolMenu.tsx:
```diff
@@ -44,6 +44,7 @@ function formatBytes(bytes: number, decimals = 2) {
     return `${parseFloat((bytes / k ** i).toFixed(dm))} ${sizes[i]}`;
 }
 
+// NOTE: When changing this name make sure that the "scripts/symbolicate-profile.ts" script is still working!
```
```suggestion
// WARNING: When changing this name make sure that the "scripts/symbolicate-profile.ts" script is still working!
```

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+const REPO = 'App';
+
+function getWorkflowRunArtifact() {
+    const artefactName = `${argsMap.platform}-sourcemap-${appVersion}`;
```
```suggestion
    const artifactName = `${argsMap.platform}-sourcemap-${appVersion}`;
```

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+// #region Utility functions
+// We need the token for the download step
+const octokit = new Octokit({auth: githubToken});
+const OWNER = 'Expensify';
```
Use constants from `@github/libs/CONST`

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+const OWNER = 'Expensify';
+const REPO = 'App';
+
+function getWorkflowRunArtifact() {
```
Let's copy the eslint config from the `.github` directory into `scripts` to allow the use of `async`/`await` for scripts

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+
+// #region Utility functions
+// We need the token for the download step
+const octokit = new Octokit({auth: githubToken});
```
*suggestion:* Update this script to use `GithubUtils.octokit` instead. You'd need to update the internals slightly in order to [supply the token from a source other than GitHub Actions inputs](https://github.com/Expensify/App/blob/1f54177b08daf1b3783ccbaeadd8a9e188e7c0c6/.github/libs/GithubUtils.ts#L74), but then I think this would be simpler and more consistent.

For example, instead of using `request`, you can use `octokit.downloadArtifact`.

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+}
+
+const dirName = '.sourcemaps';
+const sourcemapDir = path.join(process.cwd(), dirName);
```
NABish but I think this line will make it so this script only works from the root directory of App. Use this to make it slightly more robust:

```ts
const sourcemapDir = path.resolve(__dirname, '../.sourcemaps');
```

hannojg replied:
excellent idea!

roryabraham commented on scripts/symbolicate-profile.ts:
```diff
@@ -0,0 +1,208 @@
+
+// Symbolicate using the downloaded source map
+function symbolicateProfile() {
+    const command = `npx react-native-release-profiler --local ${argsMap.profile} --sourcemap-path ${localSourceMapPath}`;
```
Suggestion, add some logs and error handling:

```ts
function symbolicateProfile() {
    Logger.info(`Symbolicating profile ${argsMap.profile} with sourcemap ${localSourceMapPath}`);
    const command = `npx react-native-release-profiler --local ${argsMap.profile} --sourcemap-path ${localSourceMapPath}`;
    Logger.info(`Executing command: ${command}`);
    try {
        execSync(command, {stdio: 'inherit'});
        Logger.success('Sourcemap symbolicated!');
    } catch (err) {
        Logger.error(`Error executing command: ${command}`, err);
    }
}
```

hannojg replied:
I know that we did this in some other cases, but actually the command with execSync will:
- print to the console (stdio: inherit)
- if the command fails (the error will be printed by the tool itself), and our process will exit with the same exit code as the command invoked with execSync

So all this try catch would catch if something in the node internals fails to execute `execSync`, which is very unlikely.

So I felt the code would be leaner without that additional try catch

roryabraham replied:
oh nice, I missed the `{stdio: inherit}`, and it's actually a great trick I could've used in the past 😄

roryabraham commented:
@hoangzinh no C+ review needed here

hannojg committed:
Update scripts/symbolicate-profile.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Update src/components/ProfilingToolMenu/BaseProfilingToolMenu.tsx

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg commented:
Thanks for the review, will update the PR tmrw CET!

hannojg committed:
rename artefact -> artifact

hannojg committed:
Merge branch 'feat/script-for-symbolicating-app-profiles' of github.com:margelo/expensify-app-fork into feat/script-for-symbolicating-app-profiles

hannojg committed:
improve GithubUtils.getArtifactByName to not paginate through all artifacts, but use the name query parameter directly

hannojg committed:
refactor to allow for setup with token + add getArtifactDownloadURL helper fct

hannojg committed:
allow async/await in scripts

hannojg committed:
use GithubUtils

hannojg committed:
re=compile gh actions

hannojg committed:
comment

hannojg committed:
setup tsconfig-paths + tsconfig using it in ./scripts to use paths

hannojg committed:
run prettier

hannojg committed:
use region comments

hannojg committed:
make script executable from any location within the project

hannojg commented:
Okay, so I refactored it to use `GithubUtils`.
`GithubUtils` is imported from `@github`, which is a path alias we configured in the `tsconfig.json`. We execute the script with ts-node, which is using node;'s module strategy - this means at runtime it can't find those `@` imports.

To fix this I use [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths), and added a `tsconfig.json` inside the `./scripts` directory, so it uses `tsconfig-paths`.

hannojg commented:
Hm, I ran `npm run gh-actions-build`, but it seems that the script in the CI applies a different formatting to those files:

![Screenshot 2024-06-20 at 10 50 51](https://github.com/Expensify/App/assets/16821682/86fe4214-a8fa-403a-a640-f5a28a258839)

Not sure what to do about it

hannojg commented:
Also the typescript check fails complaining about newly added .js files - I didn't add any new js files, its just that I recompiled the GH actions.

Will take this off hold since the first PR was merged and this PR is ready for a new round of review.

roryabraham reviewed (commented):
Looks like you need to fix `Validate GitHub Actions` by rebuilding GitHub actions.

roryabraham commented on package.json:
```diff
@@ -333,5 +335,6 @@
     "npm": "10.7.0"
-  }
+  },
+  "packageManager": "yarn@1.22.22+sha512.a6b2f7906b721bba3d67d4aff083df04dad64c399707841b7acf00f6b133b7ac24255f2652fa22ae3534329dc6180534e98d17432037ff6fd140556e2bb3137e"
```
This change looks like it snuck in by mistake - can we please remove it?

hannojg replied:
ah, thanks for catching that, removed it

hannojg committed:
remove unwanted changes

hannojg committed:
ran gh-actions-build

roryabraham commented on .github/libs/GithubUtils.ts:
```diff
@@ -521,12 +529,29 @@ class GithubUtils {
+    /**
+     * Returns a single artifact by name. If none is found, it returns undefined.
+     */
     static getArtifactByName(artefactName: string): Promise<OctokitArtifact | undefined> {
```
```suggestion
    static getArtifactByName(artifactName: string): Promise<OctokitArtifact | undefined> {
```

roryabraham commented on .gitignore:
```diff
@@ -130,3 +130,6 @@ config/webpack/*.pem
 web-build/
+
+# Storage location for downloaded app source maps (see scripts/symbolicate-profile.ts)
+.sourcemaps/
```
NAB but there's no empty line at EOF:

<img width="211" alt="image" src="https://github.com/Expensify/App/assets/47436092/9a2c15c0-01e0-4a9e-9b3e-0df101baeadb">

roryabraham reviewed (changes requested):
one last thing

hannojg committed:
add EOF

roryabraham commented:
typecheck failure is a false alarm - we added a new `.eslintrc.js` file, and it's complaining about that. We can just ignore it - `tsc` is passing fine

hannojg committed:
fix parameter name spelling

hannojg committed:
jsdoc comment

hannojg committed:
ran gh-actions-build

roryabraham commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

roryabraham commented:
not an emergency - see my last comment

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9619839966) to staging by https://github.com/roryabraham in version: 9.0.1-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9668528895) to production by https://github.com/yuwenmemon in version: 9.0.1-19 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
